src/epm.rs

```diff
@@ -62,6 +72,112 @@ impl std::fmt::Display for EpmConstant {
 	}
 }
 
+#[derive(Debug)]
+pub struct State {
+	voters: Voters,
+	voters_by_stake: BTreeMap<VoteWeight, usize>,
+}
+
+impl State {
+	fn len(&self) -> usize {
+		self.voters_by_stake.len()
+	}
+
+	fn to_voters(&self) -> Voters {
+		self.voters.clone()
+	}
+}
+
+/// Represent voters that may be trimmed
+///
+/// The trimming works by removing the voter with the least amount of stake.
+///
+/// It's using an internal `BTreeMap` to determine which voter to remove next
+/// and the voters Vec can't be sorted because the EPM pallet will index into it
+/// when checking the solution.
+#[derive(Debug)]
+pub struct TrimmedVoters<T> {
+	state: State,
+	_marker: PhantomData<T>,
+}
+
+impl<T> TrimmedVoters<T>
+where
+	T: MinerConfig<AccountId = AccountId, MaxVotesPerVoter = static_types::MaxVotesPerVoter>
+		+ Send
+		+ Sync
+		+ 'static,
+	T::Solution: Send,
+{
+	/// Create a new `TrimmedVotes`.
+	pub async fn new(mut voters: Voters, desired_targets: u32) -> Result<Self, Error> {
+		let mut voters_by_stake = BTreeMap::new();
+
+		for (idx, (_voter, stake, _supports)) in voters.iter().enumerate() {
+			voters_by_stake.insert(*stake, idx);
+		}
+
+		while let Some((_, idx)) = voters_by_stake.pop_first() {
+			let rm = voters[idx].0.clone();
+
+			let mut targets = BTreeSet::new();
+			let active_voters =
+				voters_by_stake.len().try_into().expect("Voters must be < u32::MAX");
+
+			// Remove votes for an account.
+			for (_voter, _stake, supports) in &mut voters {
+				supports.retain(|a| a != &rm);
+				targets.extend(supports);
+			}
+
+			let desired_targets = desired_targets;
+			let targets = targets.len() as u32;
+
+			let est_weight: Weight = tokio::task::spawn_blocking(move || {
```

Comment:
decided to add this in the constructor/new because it should only be needed to do once